docs/source/guides/data.rst

```diff
@@ -0,0 +1,562 @@
+
+.. code-block:: python
+
+    model = LitMNIST()
+    trainer = Trainer()
+    trainer.fit(model)
+
+3. LightningDataModules (recommended)
+-------------------------------------
+Defining free-floating dataloaders, splits, download instructions, and such can get messy.
+We recommend grouping the full definition of the datasets into a :class:`~pytorch_lightning.core.datamodule.LightningDataModule` which includes:
+
+- Download instructions
+- Processing instructions
+- Split instructions
+- Train dataloader
+- Val dataloader(s)
+- Test dataloader(s)
+
+.. testcode::
+
+    class MyDataModule(LightningDataModule):
+
+        def __init__(self):
+            super().__init__()
+            self.train_dims = None
+            self.vocab_size = 0
+
+        def prepare_data(self):
+            # called only on rank 0 process
+            download_dataset()
+            tokenize()
+            build_vocab()
+
+        def setup(self, stage: Optional[str] = None):
+            # called on every process
+            vocab = load_vocab()
+            self.vocab_size = len(vocab)
+
+            self.train, self.val, self.test = load_datasets()
+            self.train_dims = self.train.next_batch.size()
+
+        def train_dataloader(self):
+            transforms = ...
+            return DataLoader(self.train, batch_size=64)
+
+        def val_dataloader(self):
+            transforms = ...
+            return DataLoader(self.val, batch_size=64)
+
+        def test_dataloader(self):
+            transforms = ...
+            return DataLoader(self.test, batch_size=64)
+
+        def predict_dataloader(self):
+            transforms = ...
+            return DataLoader(self.predict, batch_size=64)
+
+DataModules are easier to re-use compared to pure :class:`~torch.utils.data.Dataset` definitions.
+
+.. code-block:: python
+
+    # use an MNIST dataset
+    mnist_dm = MNISTDatamodule()
+    model = LitModel(num_classes=mnist_dm.num_classes)
+    trainer.fit(model, mnist_dm)
+
+    # or other datasets with the same model
+    imagenet_dm = ImagenetDatamodule()
+    model = LitModel(num_classes=imagenet_dm.num_classes)
+    trainer.fit(model, imagenet_dm)
+
+.. note:: :meth:`~pytorch_lightning.core.datamodule.LightningDataModule.prepare_data` is called on only one GPU in distributed training (automatically)
+.. note:: :meth:`~pytorch_lightning.core.datamodule.LightningDataModule.setup` is called on every GPU (automatically)
+
+Read :ref:`this <datamodules>` for more details on LightningDataModules.
+
+---------------
+
+*****************************************
+Dynamically Configuring Models using Data
+*****************************************
+
+When your models need to know about the data, it's best to process the data before passing it to the model.
+
+.. code-block:: python
+
+    # init dm AND call the processing manually
+    dm = ImagenetDataModule()
+    dm.prepare_data()
+    dm.setup()
```

Comment:
in this case, the guarantee that this is called on every process is no longer satsified, right?

Reply:
Yes, I don't think it is. It could be if we were wrapping the setup function and checking the rank.

Reply:
should i add a comment?